### DISCUSSION TITLE
Admin console context path issues behind rewriting nginx proxy using --hostname-path (env KC_HOSTNAME_PATH)

### DISCUSSION THREAD

**Discussion by poconnor-lab49** [keycloak/keycloak#11546] [Q&A]
We're trying to migrate from Keycloak 16.1.1 to 17.x and are having some issues reaching the admin console. 

The initial request works, but all the subsequent requests fail with a 404 since they don't seem to use the context path. For example, https://example.com/auth-test loads the "Welcome to Keycloak" HTML, but the next request to https://example.com/resources/5zn7g/common/keycloak/node_modules/patternfly/dist/css/patternfly.css which returns a 404.

We're building a custom image with no environment variable overrides except `KC_METRICS_ENABLE=true` and `KC_DB=postgres`.  I don't think I should need to set `--http-relative-path` or `KC_HTTP_RELATIVE_PATH` since nginx is handling the routing and is rewriting  `/auth-test(/|$)(.*)`  to `/$2`.

Keycloak is deployed with environment variables like
```sh
      KC_PROXY:                     edge
      KC_DB:                        postgres
      KC_DB_URL_HOST:               keycloak-test-db-postgresql
      KC_DB_URL_PORT:               5432
      KC_METRICS_ENABLED:           true
      KC_HOSTNAME:                  example.com
      KC_HOSTNAME_PATH:             auth-test
      KC_HOSTNAME_STRICT:           false
      KC_HOSTNAME_STRICT_HTTPS:     false
      KC_HTTP_ENABLED:              true
```

If I navigate directly to `https://example.com/auth-test/admin/` the interface loads correctly, but all the links are still broken.

Navigating to `https://example.com/auth-test/realms/master/account/#/` loads a broken page, but the `Sign in` button works.

Am I using this wrong, or is there some other environment variable I should be using instead?  Do I need to set a quarkus property directly?

The reason I'm asking is I want to be able to deploy multiple keycloak instances behind the same proxy at different paths so that I can keep a stable instance running for developers while I test changes.

**Comment by poconnor-lab49** [comment]
Seems like its this bug maybe? https://github.com/keycloak/keycloak/issues/10321

**Reply by Brandt1930** [reply]
Hi, did you manage to solve it? I'm stuck on the same issue.

**Reply by poconnor-lab49** [reply]
Unfortunately, no.  TL;DR I got moved to another project and didn't need that functionality any more.

I don't know much about keycloak, so take this with a grain of salt.  I *think* I was facing two problems:

- a case sensitive check for the `X-Forwarded-*` headers, which envoy was downcasing so they were ignored by keycloak.  This is maybe an upstream issue with the headers library? I'm pretty sure HTTP header field names are supposed to be case insensitive.
- `QuarkusWelcomeResource` was not adding the context path to the URIs when proxy mode is `edge`. So `theme/keycloak/welcome/index.ftl` had broken URLs.  You could see the issue right away in broken image links.

Not sure if the same issue exists for keycloak 18 and later.  I know there is a new default admin console so maybe this is fixed now?  

Since no one ever confirmed that this should have worked the way I described, I didn't bother opening a bug.  I'm still not sure if maybe I didn't just miss something (shrug).

EDIT: Strike case sensitive headers as an issue.  I got curious, so I built locally and modified to use lowercase headers https://github.com/keycloak/keycloak/blob/main/quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/ProxyDistTest.java#L75 and it seems fine to me :)

**Reply by darius-m** [reply]
I have also run into a similar issue in Keycloak 19.0.1, and the admin interface did not properly load when setting the `hostname-path` configuration option. In our case, we managed to get it working by setting `http-relative-path` to the path  (e.g. `/auth`) instead, and left `hostname-path` unset.

**Reply by Brandt1930** [reply]
Thanks guys. I had the issue that the admin-console was not available changing to a the old relative path `/auth`. Keycloak runs behind a reverse proxy and hovering over the "Administration Console" Link on the Welcome Page I always only saw `<hostname>/master/console` not `/auth/master/console`. But in the current nightly builds there is an option to set the admin-console-path as well as the hostname-path (#13524) . So I set  `KC_HOSTNAME_URL: https://<hostname>/auth/`and `KC_HOSTNAME_ADMIN_URL: https://<hostname>/auth/` and now it works :)
